File: Rummikub.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Scanner;

public class Rummikub {
    // Might need to return 2-4 player's racks
    public static HashMap<String, ArrayList<Tile>> allocateTilesAtStart(int numPlayers) {
        HashMap<String, ArrayList<Tile>> tileLists = new HashMap<String, ArrayList<Tile>>();
        
        ArrayList<Tile> poolTiles = putTilesInPool();
        ArrayList<Tile> p1Tiles = new ArrayList<Tile>();
        ArrayList<Tile> p2Tiles = new ArrayList<Tile>();
        ArrayList<Tile> p3Tiles = new ArrayList<Tile>();
        ArrayList<Tile> p4Tiles = new ArrayList<Tile>();

        // Allocate 14 random tiles to each player.
        for (int i = 0; i < 14 * numPlayers; ++i) {
            Tile t1 = poolTiles.get(getRandom(0, poolTiles.size() - 1));
            
            switch(i / 14) {
                case 0:
                    p1Tiles.add(t1);
                    break;
                case 1:
                    p2Tiles.add(t1);
                    break;
                case 2:
                    p3Tiles.add(t1);
                    break;
                case 3:
                    p4Tiles.add(t1);
                    break;
                default:
                    break;
            }

            poolTiles.remove(t1);
        }

        // Only return the non-empty lists (always includes the 1st 3 lists).
        tileLists.put("Pool tiles", poolTiles);
        if (p1Tiles.size() != 0) { tileLists.put("Player 1 tiles", p1Tiles); };
        if (p2Tiles.size() != 0) { tileLists.put("Player 2 tiles", p2Tiles); };
        if (p3Tiles.size() != 0) { tileLists.put("Player 3 tiles", p3Tiles); };
        if (p4Tiles.size() != 0) { tileLists.put("Player 4 tiles", p4Tiles); };

        return tileLists;
    }

    public static boolean checkBoardLength(ArrayList<ArrayList<Tile>> sequences) {
        for (int i = 0; i < sequences.size(); ++i) {
            if (sequences.get(i).size() < 3) {
                System.out.println("Sequence " + (i + 1) + " is not long enough");
                return false;
            }
        }

        return true;
    }

    public static boolean checkBoardSequences(ArrayList<ArrayList<Tile>> sequences) {
        /* Only the 1st group- and run-related errors are captured for simplicity.
         * Also, each element must be specifically initialised as an object.
         */
        ErrorMessage[] groupRunErrors = new ErrorMessage[sequences.size()];
        for (int i = 0; i < groupRunErrors.length; ++i) {
            groupRunErrors[i] = new ErrorMessage();
        }
        
        // Groups
        for (int i = 0; i < sequences.size(); ++i) {
            ArrayList<Tile> currSequence = sequences.get(i);
            for (int j = 0; j < currSequence.size(); ++j) {
                // Make sure colours aren't duplicated
                ArrayList<String> usedColours = new ArrayList<String>();
                usedColours.add(currSequence.get(j).getColour());

                if (j == 4) { // Group length cannot exceed 4 (incl. jokers)
                    groupRunErrors[i].setGroupError("Cannot have a group of 5 or more tiles");
                    break;
                }
    
                // Colours must not be used more than once, bar jokers.
                if ((usedColours.contains(currSequence.get(j).getColour())) &&
                    (currSequence.get(j).getNumber() != 0)) {
                    groupRunErrors[i].setGroupError("Cannot have 2 non-jokers of the same colour per group");
                    break;
                }

                // Numbers are only allowed to differ in the context of a joker.
                if (j <= currSequence.size() - 1) { // Check for out of bounds access
                    if ((currSequence.get(j).getNumber() != currSequence.get(j + 1).getNumber()) &&
                        ((currSequence.get(j).getNumber() != 0) || (currSequence.get(j + 1).getNumber() != 0))) {
                        groupRunErrors[i].setGroupError("Cannot have 2 non-jokers with different numbers per group");
                        break;
                    }
                }
            }
        }

        // Runs
        for (int i = 0; i < sequences.size(); ++i) {
            ArrayList<Tile> currSequence = sequences.get(i);
            for (int j = 0; j < currSequence.size(); ++j) {
                // Check for out of bounds numbers and where a joker is in between 2 offending tiles
                if ((j > 0) && (j < currSequence.size() - 1)) {
                    switch (currSequence.get(j).getNumber()) {
                        case 0: // Jokers, noting that there may be 2 of them in a row
                            if (!(currSequence.get(j-1).getColour().equals(currSequence.get(j+1).getColour())) ||
                                 (currSequence.get(j-1).getColour().equals(currSequence.get(j+2).getColour())) ||
                                 (currSequence.get(j-2).getColour().equals(currSequence.get(j+1).getColour()))) {
                                groupRunErrors[i].setRunError("Colour mismatch overlapping with a joker/s");
                                break;
                            }

                            if (!((currSequence.get(j-1).getNumber() == currSequence.get(j+1).getNumber()-2) ||
                                  (currSequence.get(j-2).getNumber() == currSequence.get(j+1).getNumber()-3) ||
                                  (currSequence.get(j-1).getNumber() == currSequence.get(j+2).getNumber()-3))) {
                                groupRunErrors[i].setRunError("Number mismatch overlapping with a joker/s");
                                break;
                            }
                        case 1:
                            groupRunErrors[i].setRunError("Cannot have a run with a 1 as a non-first tile");
                            break;
                        case 13:
                            groupRunErrors[i].setRunError("Cannot have a run with a 13 as a non-last tile");
                            break;
                        default:
                            break;
                    }

                    // Don't capture multiple errors per group (or run) check per sequence.
                    if (!groupRunErrors[i].getRunError().equals("")) { continue; }
                }
                else if (j < currSequence.size() - 1) { // Either of the above conditions would have done.
                    if ((currSequence.get(j+1).getNumber() != 0) &&
                        (currSequence.get(j+1).getNumber() != currSequence.get(j).getNumber() + 1)) {
                       groupRunErrors[i].setRunError("Cannot have a broken run sequence");
                    }
                    
                    // Don't capture multiple errors per group (or run) check per sequence.
                    if (!groupRunErrors[i].getRunError().equals("")) { continue; }

                    if (!currSequence.get(j+1).getColour().equals(currSequence.get(j).getColour())) {
                       groupRunErrors[i].setRunError("Cannot have multiple colours per run sequence");
                    }
                }
            }
        }
        
        int numFalse = 0;

        for (int i = 0; i < groupRunErrors.length; ++i) { // Count number of, and print errors.
            // A sequence should never be both a group and a run, but might be neither.
            if (!((groupRunErrors[i].getGroupError().equals("")) ^ (groupRunErrors[i].getRunError().equals("")))) {
                numFalse++;

                System.out.println("Sequence " + (i + 1) + " errors");
                System.out.println(groupRunErrors[i].getGroupError());
                System.out.println(groupRunErrors[i].getRunError());
                System.out.println();
            }        
        }

        return numFalse == 0;
    }

    /* A group is a 3- or 4-tile row of same-numbered tiles but different colours.
     * A run is a 3-tile (or more) row of same-coloured tiles in ascending order.
     */
    public static boolean checkBoardValidity(ArrayList<ArrayList<Tile>> sequences) {
        // Separate from other console output regardless of what might/mightn't be outputted.
        System.out.println();
        
        // TODO: Needs reworking
        boolean lengthOK = checkBoardLength(sequences); // All sequences have at least 3 tiles
        boolean groupRunOK = checkBoardSequences(sequences); // All sequences are either groups or runs

        // The 2nd function checks for run bounds.
        return lengthOK && groupRunOK;
    }

    // Only checks against runs.
    public static boolean checkRunBounds(ArrayList<Tile> currSequence, int currNum) {
        int currSize = currSequence.size();

        for (int j = 0; j < currSequence.size(); ++j) {
            if ((j > 0) && (currSequence.get(j).getNumber() == 1) && (currSequence.get(0).getNumber() != 1)) {
                System.out.print("Sequence " + (currNum + 1) + ": The number 1 must only be on ");
                System.out.println("the first tile, unless part of a group");
                return false;
            }

            if ((j < currSize - 1) && (currSequence.get(j).getNumber() == 13) && (currSequence.get(currSize - 1).getNumber() != 13)) {
                System.out.print("Sequence " + (currNum + 1) + ": The number 13 must only be on ");
                System.out.println("the last tile, unless part of a group");
                return false;
            }
        }

        return true;
    }

    public static boolean emptyRack(HashMap<String, ArrayList<Tile>> tiles, int currPlayer) {
        // The keys are (Pool | Player 1-4) tiles
        switch(currPlayer){
            case 1:
                return tiles.get("Player 1 tiles").size() == 0;
            case 2:
                return tiles.get("Player 2 tiles").size() == 0;
            case 3:
                return tiles.get("Player 3 tiles").size() == 0;
            case 4:
                return tiles.get("Player 4 tiles").size() == 0;
            default: // Assume that some invalid player number has no rack
                return false;
        }
    }
    
    // Used for selecting both the starting player and tiles from pool
    public static int getRandom(int min, int max) {
        /* 
         * Since Math.random() returns a real number in range [0, 1), the multiplier
         * acts as the number of possible numbers to choose from. As the list of possible
         * numbers is contiguous (with both parameters inclusive), the minimum value
         * can be set by adding it to the result. Therefore, a number in range [min, max]
         * gets returned.
         */
        return (int)(Math.random() * (max - min + 1)) + min;
    }

    public static void printBoard (ArrayList<ArrayList<Tile>> sequences) {
        System.out.println();
        
        int i = 0;
        for (i = 0; i < sequences.size(); ++i) {           
            System.out.print("Sequence " + (i + 1) + ": <");
         
            for (int j = 0; j < sequences.get(i).size(); j++) {
                System.out.print(sequences.get(i).get(j).toString());
                if (j < sequences.get(i).size() - 1) {
                    System.out.print(", ");
                }
                else {
                    System.out.println(">");
                }
            }
        }

        // Simple way of checking the size. Obviously requires pre-loop initalisation
        if (i == 0) { System.out.println("Board is empty"); }
    }

    public static void printPlayerTiles (HashMap<String, ArrayList<Tile>> tiles, int currPlayer) {        
        // The keys are (Pool | Player 1-4) tiles
        ArrayList<Tile> currTiles = new ArrayList<Tile>();
        
        switch(currPlayer){
            case 1:
                currTiles = tiles.get("Player 1 tiles");
                break;
            case 2:
                currTiles = tiles.get("Player 2 tiles");
                break;
            case 3:
                currTiles = tiles.get("Player 3 tiles");
                break;
            case 4:
                currTiles = tiles.get("Player 4 tiles");
                break;
            default: // Assume that some invalid player number has no rack
                break;
        }
        
        // TODO: Sort by colour (numbers are not yet touched).
        // currTiles.sort(null);

        System.out.println("\nPlayer " + currPlayer + " has the following tiles:");
        for (int i = 0; i < currTiles.size(); ++i) {
            System.out.print(currTiles.get(i) + (i < currTiles.size() - 1 ? ", " : ""));
        }
        System.out.println();
    }

    public static void printUserCommands(int numPoolTiles, int numSequences) {
        System.out.println("\nChoose from one of the following commands (case insensitive):");
        
        System.out.println("Abort: Revert game state back to the last move");
        System.out.println("Finish: Finish current move and save current game state.");
        
        if (numSequences > 0) {
            System.out.print("Split: Split a chosen sequence at the specified turn. ");
            System.out.println("Note: Currently assumes there will be no gaps left in the middle");
        }
        
        if (numPoolTiles > 0) {
            System.out.println("Pool: Grab 1 tile from the pool (forfeits turn)");
        }

        // Extra blank line that need not be moved around as a result of building this method.
        System.out.println();
    }

    public static ArrayList<Tile> putTilesInPool() {
        ArrayList<Tile> poolTiles = new ArrayList<Tile>();
        final String[] COLOURS = {"Blue", "Orange", "Black", "Red"};

        // Add non-jokers
        for (int i = 1; i <= 13; ++i) {
            for (String c1 : COLOURS) {
                for (int j = 0; j < 2; ++j) {
                    poolTiles.add(new Tile(c1, i));
                }
            }
        }

        // Add jokers
        poolTiles.add(new Tile("Black"));
        poolTiles.add(new Tile("Red"));
       
        return poolTiles;
    }

    public static HashMap<String, ArrayList<Tile>> randomPoolTile(HashMap<String, ArrayList<Tile>> tileLists, int currPlayer) {
        if (tileLists.get("Pool tiles").size() == 0) {
            System.out.println("Sorry, no more tiles left. Your move has still been forfeited");
        }
        
        Tile t1 = tileLists.get("Pool tiles").get(getRandom(0, tileLists.get("Pool tiles").size() - 1));
        
        tileLists.get("Player " + currPlayer + " tiles").add(t1);
        tileLists.get("Pool tiles").remove(t1);
        
        return tileLists;
    }

    public static int setNumPlayers(Scanner scnr) {
        System.out.println("Welcome to Rummikub. Select between 2-4 players");
        int numPlayers = scnr.nextInt();
        
        while ((numPlayers < 2) || (numPlayers > 4)) {
            System.out.println("Error: Must select between 2-4 players");
            numPlayers = scnr.nextInt();
        }

        return numPlayers;
    }

    public static int setStartPlayer(Scanner scnr, int numPlayers) {
        System.out.println("Choose a starting player (or 5 for random)");
        int startPlayer = scnr.nextInt();
        
        // Escape to the next line, as this will be used once again for the next user input.
        scnr.nextLine();
        
        /*
         * Accepted values are in range [1, numPlayers] and 5. Cannot simply check for
         * 2 sides of a range because numPlayers is in range [2, 4] i.e. not always 4.
         */
        while ((startPlayer < 1) || ((startPlayer > numPlayers) && (startPlayer != 5))) {
            System.out.println("Error: Must select a valid number (or 5 for random)");
            startPlayer = scnr.nextInt();
            scnr.nextLine(); // Escape to the next line, see main for more details.
        }

        if (startPlayer == 5) {// Inform of starting player, if randomly selected
            startPlayer = getRandom(1, numPlayers);
            System.out.println("Player " + startPlayer + " is starting");
        }
        
        return startPlayer;
    }

    // Currently assumes that the sequence will be split so as to not leave empty spaces
    public static ArrayList<ArrayList<Tile>> splitSequence(ArrayList<ArrayList<Tile>> sequences, Scanner scnr) {
        boolean validInput = false;
        int sequenceNumber = 0;
        int splittingPosition = 0;

        while (!validInput) {
            // Declare false once rather than declare true multiple times
            validInput = true;
        
            // Defined as the first number to go into the new array
            System.out.println("\nEnter the sequence number followed by the number of the 1st element to be moved");
            sequenceNumber = scnr.nextInt();
            splittingPosition = scnr.nextInt();
            
            // Indices are [1, n] in human readable format, stored as [0, n)
            if ((sequenceNumber < 1) || (sequenceNumber > sequences.size())) {
                System.out.println("Error: Sequence number not found");
                validInput = false;
                
                /* Not only would splittingPosition not have a valid value for
                 * a non-existent sequence, continuing with the iteration would
                 * also cause an out of bounds exception
                 */
                continue;
            }

            if (splittingPosition < 1) {
                System.out.println("Error: Cannot split at a non-positive position");
                validInput = false;
            }
            else if (splittingPosition == 1) {
                System.out.println("Error: Cannot split an entire sequence");
                validInput = false;
            }
            else if (splittingPosition > sequences.get(sequenceNumber - 1).size()) {
                System.out.println("Error: Sequence " + (sequenceNumber - 1) + " contains fewer than " + splittingPosition + " numbers");
                validInput = false;
            }
        }

        /* Copy the tiles to a new ArrayList, then put it after the original sequence
         * (everything gets shifted by 1). The sequence number must be decremented
         * to convert from [1, n] to [0, n) style, then incremented so that the new
         * sequence comes directly after. However, doing so in the loop will affect
         * the results prematurely.
         */
        sequenceNumber--;
        splittingPosition--;
        
        /* Do things in reverse order to avoid messing up the indices and double loops.
         * Instead of adding at the next available index, must add at the beginning to
         * preserve order. Anything pre-existing will be shifted by 1.
         */
        ArrayList<Tile> newSequence = new ArrayList<Tile>();
        for (int i = sequences.get(sequenceNumber).size() - 1; i >= splittingPosition; --i) {
            newSequence.add(0, sequences.get(sequenceNumber).get(i));
            sequences.get(sequenceNumber).remove(i);
        }
        sequences.add(++sequenceNumber, newSequence);
        
        scnr.nextLine(); // Escape in preparation for the next command
        return sequences;
    }

    public static void main(String[] args) {
        ArrayList<ArrayList<Tile>> sequences = new ArrayList<ArrayList<Tile>>();
        Scanner scnr = new Scanner(System.in);

        // 2nd parameter in setStartPlayer() allows a random player to start
        int numPlayers = setNumPlayers(scnr);
        int currPlayer = setStartPlayer(scnr, numPlayers); // Initialised then cycles through.
        
        /* The keys are (Pool | Player 1-4) tiles. Cannot include the sequences list as it is
         * 2D (must be able to store which seqeuence a tile belongs to)
         */
        HashMap<String, ArrayList<Tile>> tileLists = allocateTilesAtStart(numPlayers);
        
        /* Each player must put down 30 or more points worth of their
         * own tiles (without touching any other tiles).
         */
        HashMap<String, Boolean> firstMoves = new HashMap<String, Boolean>();
        for (int i = 1; i <= numPlayers; ++i) {
            firstMoves.put("Player " + i, false);
        }

        sequences.add(tileLists.get("Player 1 tiles"));

        boolean gameFinished = false;
        while (!gameFinished) {
            printBoard(sequences);
            printPlayerTiles(tileLists, currPlayer);
            printUserCommands(tileLists.get("Pool tiles").size(), sequences.size());
            
            boolean validInput = false;
            while (!validInput) {
                /* Declare false once rather than declare true multiple times.
                 * Note that this represents whether the command is recognised,
                 * as well as if it could be executed given the current game state.
                 * Printing the game state and/or user commands would be too much
                 * directly following invalid input.
                 */
                validInput = true;

                // Store the game state as of the most recent move in case of desire to abort move.
                final ArrayList<ArrayList<Tile>> SEQUENCES_LAST_VERSION = sequences;
                final HashMap<String, ArrayList<Tile>> TILELISTS_LAST_VERSION = tileLists;

                String userInput = scnr.nextLine();
                switch (userInput.toUpperCase()){
                    // case "ADD":
                    case "ABORT": // Revert game state to original state.
                        sequences = SEQUENCES_LAST_VERSION;
                        tileLists = TILELISTS_LAST_VERSION;

                        System.out.println("Move successfully aborted.");
                        break;
                    case "FINISH": // Finish current move. The inner while-loop will 
                        if (checkBoardValidity(sequences)) {
                            System.out.println("Move successfully finished.");
                        }
                        else {
                            validInput = false;
                            System.out.println("Please use the above commands to any mistakes and try again");
                        }

                        break;
                    case "SPLIT": // Should only appear where upon execution, the if-statement would be true.
                        if (sequences.size() > 0) {
                            sequences = splitSequence(sequences, scnr);
                        }
                        else {
                            validInput = false;
                            System.out.println("Error: No sequences to split.");
                        }

                        break;
                    case "POOL": // Should only appear where upon execution, the if-statement would be true.
                        if (tileLists.get("Pool tiles").size() > 0) {
                            tileLists = randomPoolTile(tileLists, currPlayer);
                        }
                        else {
                            validInput = false;
                            System.out.println("Error: No tiles left in the pool.");
                        }
                        
                        break;
                    // case "SWAP":
                    default:
                        validInput = false;
                        System.out.println("Error: Command unrecognised");
                        break;
                }
            }

            if (emptyRack(tileLists, currPlayer)) {
                gameFinished = true;
            }
            else {
                // Increment player number (or set to 1 if maxed out)
                currPlayer = (currPlayer < numPlayers) ? currPlayer + 1 : 1;
            }
        }

        System.out.println("\nPlayer " + currPlayer + " won!");
    }
}

File: Rummikub.py
import math, random # Floor function and random number, respectively
from Objects import * # Class definitions
    
def allocateTiles(numPlayers):
    pool, player1Rack, player2Rack, player3Rack, player4Rack = [], [], [], [], []
    
    # Add non-jokers
    for i in range(1, 13+1): # Tile numbers are in range[1, 13] but end index is exclusive
        for c1 in ["Blue", "Orange", "Black", "Red"]:
            for j in range(2):
                pool.append(Tile(c1, i))

    # # Add jokers
    # pool.append(Tile("Black"))
    # pool.append(Tile("Red"))

    for i in range(0, 14 * numPlayers): # Add 14 tiles to each player by random assignment
        newTile = pool.pop(random.randint(0, len(pool) - 1)) # Analogous to Ctrl+X
        
        # By cycling between the number of tiles (as opposed to that of the players), we avoid assigning tiles to extra players (and potentially in an uneven fashion)
        match math.floor(i / 14):
            case 0:
                player1Rack.append(newTile)
            case 1:
                player2Rack.append(newTile)
            case 2:
                player3Rack.append(newTile)
            case 3:
                player4Rack.append(newTile)

    return pool, [player1Rack, player2Rack, player3Rack, player4Rack] # Easier if the racks are in one array

# Returns true if the sequence contains 3 or 4 tiles of varying colour (containing the same numbers and/or jokers)
def checkGroup(sequence):
    if len(sequence) > 4: # A 5 (or more)-tile sequence is guaranteed to repeat colours, irrespective of jokers
        return False

    # Get separate lists of the numbers and colours found among the non-jokers
    (numbersList, coloursList) = [(x.getNumber(), x.getColour()) for x in sequence if x.getNumber() > 0]
    
    if len(set(numbersList)) > 1: # Must only find one unique number besides jokers
        return False    
    if len(coloursList) != len(set(coloursList)): # Any colour repeats will only show up in the 1st list.
        return False
    
    return True # If all checks have passed

# Returns true if the sequence is strictly increasing and of the same colour
def checkRun(sequence):
    # Get separate lists of the numbers and colours found among the non-jokers
    (numbersList, coloursList) = [(x.getNumber(), x.getColour()) for x in sequence if x.getNumber() > 0]

    if len(set(coloursList)) > 1: # Must only find one unique colour besides jokers
        return False    
    if len(numbersList) != len(set(numbersList)): # Any number repeats will only show up in the 1st list.
        return False

    # Check if the non-jokers are in the right order
    non_zeros = [x for x in numbersList if x != 0]
    if non_zeros != sorted(non_zeros):
        return False
    
    # Check if there are missing numbers not covered by jokers (may not be possible using list manipulations/comprehensions from above)
    currNumber = numbersList[0]
    for i in range(1, len(numbersList)):
        if numbersList[i] not in (numbersList[i-1] + 1, 0): # Perform 2 comparisons in one go
            return False
        currNumber += 1
    
    # If duplicate numbers have been caught earlier, there has to be a joker at at least one edge index
    if any([1, 13]) in numbersList[1:-1]:
        return False
    
    return True # If all checks have passed


def getNumPlayers():
    numPlayers = None
    
    try:
        numPlayers = int(input("Please enter the number of players (2-4)\n"))
    except Exception as e: # Wrong type
        numPlayers = 0 # Ensure wrong input doesn't get treated as valid
    
    return numPlayers

def getStartPlayer():
    startPlayer = None
    
    try:
        startPlayer = int(input("Please choose the player number (or 5 for random)\n"))
    except Exception as e: # Wrong type
        startPlayer = 0 # Ensure wrong input doesn't get treated as valid
    
    return startPlayer

def printTileList(listName, tileList):
    if len(tileList) == 0:
        print("\n" + listName, "is empty")
    else:
        print("\n" + listName + ":")
        if listName == "Board":            
            for i in range(len(tileList)):
                print("Sequence", str(i+1), end = ": [")
                for j in range(len(tileList[i])):
                    print(tileList[i][j], end = ']\n' if j == len(tileList[i]) - 1 else ", ")
        else:
            for i in range(len(tileList)):
                if i == 0:
                    print("[", end = "")
                
                print(tileList[i], end = ", " if i < len(tileList) - 1 else "]\n\n")

def printUserCommands(numPoolTiles, numSequences):
    print("Choose from one of the following commands (case insensitive):")
    print("Abort: Revert game state back to the last move")
    print("Finish: Finish current move and save current game state.")
    
    if numSequences > 0:
        # print("Split: Split a chosen sequence at the specified turn. ")
        # print("Note: Currently assumes there will be no gaps left in the middle")
        pass
    
    if numPoolTiles > 0:
        print("Pool: Grab 1 tile from the pool (forfeits turn)")

    # Extra blank line that need not be moved around as a result of building this method.
    print()   

def validBoard(board):
    # # These are the following possible errors (if a sequence is valid, only one of 2a and 2b will be triggered):
    # # 1. A sequence containing < 3 tiles
    # # 2a. A sequence is not a group (i.e. the pattern is not entirely colour-based)
    # # 2b. A sequence is not a run (i.e. the pattern is not entirely number-based)

    # errorMessageList = []
    # for index, sequence in board:
    #     # Error 1 from above
    #     if len(sequence) < 3:
    #         errorMessageList.add(ErrorMessage("Sequence", index, "has < 3 tiles"))
    #         continue
               
    #     # Error 2a, 2b from above
    #     isGroup, isRun = checkGroup(sequence), checkRun(sequence)
    #     if not(isGroup ^ isRun): # All valid sequences are either groups or runs, not neither nor both.
    #         errorMessageList.append(ErrorMessage("Sequence", index, "must strictly increase or change colours, not neither nor both"))
    #         continue

    # for em in errorMessageList: print("Error:", em)
    # return len(errorMessageList) == 0
    return True # Temporary placeholder whilst the above hasn't been finished

if __name__ == '__main__': # Main method
    print("Welcome to Rummikub!", end = " ")
    
    # Get number of players
    numPlayers = getNumPlayers()
    while (numPlayers < 2 or numPlayers > 4):
        print("Error: Number of players must be between 2 and 4.", end = " ")
        numPlayers = getNumPlayers()
    
    # Choose the starting player explicitly or randomly
    currPlayer = getStartPlayer()
    while (currPlayer < 1 or (currPlayer > numPlayers and currPlayer != 5)):
        print("Error: Player number must be between 1 and " + str(numPlayers) + " (or 5 for random)", end = " ")
        currPlayer = getStartPlayer()
    
    # Randomly assign starting player (if necessary) and print out the final decision
    if currPlayer == 5:
        currPlayer = random.randint(1, numPlayers)
    print("Player", currPlayer, "is starting")

    # Allocate tiles to the pool and all players
    pool, playerRacks = allocateTiles(numPlayers) # Again, it's easier if the player racks are all under one array
    if numPlayers < 4: # The respective lists should be empty anyway, but destroy them just in case
        del playerRacks[3]
        if numPlayers < 3:
            del playerRacks[2]

    board = [] # 2D-array of sequences
    finished = False
    while not finished: # Game-wide flow
        draftRack = playerRacks[currPlayer - 1].copy() # Copy the current rack so that it can be returned to later if need be
        poppedTiles = [] # Keep track of tiles grabbed from the pool
        
        printTileList("Board", board) # Print the board
        printTileList("Player " + str(currPlayer) + "'s rack", playerRacks[currPlayer - 1]) # Print the current rack
        while True: # Turn-wide flow
            printUserCommands(len(pool), sum(len(sequence) for sequence in board)) # Number of tiles in the pool and board, respectively
            command = input().lower() # Could also use upper(), since case insensitivity is the goal

            # If there is a simple condition like empty list that can invalidate a command, the user cannot see it nor will
            # they be able to enter it regardless. However, it's up to the user to check if the board is valid.
            match command:
                case "abort": # Discard all changes
                    pool.extend(poppedTiles) # Return any required tiles back to the pool
                    break
                case "finish": # Finalise all changes, assuming the board is in a legal state.
                    if validBoard(board):
                        playerRacks[currPlayer - 1] = draftRack.copy() # Copy the draft rack back to the working copy.
                        break
                    continue
                case "pool" if len(pool) > 0: # Randomly assign a tile from the pool if possible
                    randTile = pool.pop(random.randint(0, len(pool) - 1))
                    draftRack.append(randTile)
                    poppedTiles.append(randTile)

                    printTileList("Player " + str(currPlayer) + "'s rack", draftRack) # Print the current rack
                    continue
                case _:
                    print("Error: Command not recognised")

        if len(playerRacks[currPlayer - 1]) == 0: # Current player emptied all their tiles
            print("Player", currPlayer, "won!")
            finished = True
        else: # Next player
            currPlayer = 1 if currPlayer == numPlayers else currPlayer + 1